kf_lib_data_ingest/validation/data_validator.py

```diff
@@ -0,0 +1,435 @@
+
+        :param dict_of_dataframes: dict with filename keys and dataframe values
+        :yield: test result dicts
+        """
+        logger.info("---")
+        logger.info("Validating data values...")
+        for concept, (desc, func) in INPUT_VALIDATION.items():
+            logger.info(f"{concept} {desc}...")
+            errors = {}
+            tested = False
+            for fname, df in dict_of_dataframes.items():
+                bad_ones = []
+                if concept in df:
+                    tested = True
+                    bad_ones = list(
+                        df[concept][~df[concept].apply(func)].unique()
+                    )
+                    if bad_ones:
+                        errors[fname] = bad_ones
+
+            yield self._format_result(
+                "attribute", f"{concept} {desc}", tested, errors
+            )
+
+    def _build_graph(self, dict_of_dataframes, include_implicit=True):
+        """Construct a graph that represents the study data. Nodes are cells in the
+        data, edges are relationships between cells across rows according to the
+        designated relationship hierarchy.
+
+        :param dict_of_dataframes: dict with filename keys and dataframe values
+        :param include_implicit: whether to deduce implied connections, default True
+        :return: a graph representation of the data
+        """
+        prep = "WITH" if include_implicit else "WITHOUT"
+        logger.info(
+            f"Building node graph {prep} implied connection discovery..."
+        )
+        graph = GroupingGraph()
+        remaining_edges = set()
+
+        # ##### Add all nodes and record valid colinear pairs as edges ##### #
+
+        logger.info("Adding nodes and direct edges...")
+
+        for df in dict_of_dataframes.values():
+            for row in df.to_dict(orient="records"):
+                nodes = [(k, v) for k, v in row.items() if v != NA]
+
+                for c, n in nodes:
+                    graph.add_node((c, n))
+
+                # If a node has a direct hierarchical connection within a row, we
+                # can probably assume that indirect connections in the same row are
+                # superfluous
+                #
+                # We should probably also only connect a given start node to its
+                # hierarchically closest destination nodes
+
+                row_edges = list(combinations(nodes, 2))
+                direct = set()
+
+                # Direct edges go in right away and the source nodes get marked
+                for a, b in row_edges:
+                    if b[0] in self.HIERARCHY_DIRECT[a[0]]:
+                        graph.add_edge(a, b)
+                        direct.add(a)
+                    elif a[0] in self.HIERARCHY_DIRECT[b[0]]:
+                        graph.add_edge(b, a)
+                        direct.add(b)
+
+                # Indirect edges get grouped by distance from source node
```

Comment:
No I don't think so.

> at that point you know that the edge (a,b) is not a direct edge

Yes, but you don't know whether a has _any_ direct edge until you've looked at all of the possible edges in the row.

Reply:
I'm not sure. I was thinking that it may be sufficient for a to have _any_ direct link into the graph to disregard any potential indirect links out of it.

Reply:
No, I think you need all of the direct edges in place before you can determine what to do for the potential indirect ones.

The only alternative I can think of to what I said about about _any_ direct link out of `a` invalidating all potential indirect links out of `a` is to fully check for connectedness between `a` and `b` in the graph, which is much slower.

Reply:
In short, eliminating those few lines will increase the run time significantly.